### ISSUE TITLE
[Feature request] Add automatic python3 bindings for the library

### ISSUE THREAD

**Issue by Cameronsplaze** [LibVNC/libvncserver#502]
If you'd like to put out an incentive for implementing this feature, you can do so at https://issuehunt.io/r/LibVNC/libvncserver

**Is your feature request related to a problem? Please describe.**

I would like to use a vnc library in python, and re-writing one from scratch seems overkill. There aren't great alternatives in that language.

**Describe the solution you'd like**

By creating automatic python bindings from this library, you cover two languages with one. Plus maybe lay ground work for more later on.

**Describe alternatives you've considered**

[RealVNC](https://www.realvnc.com/en/developer/features/) is the only library with python bindings I can find, and is very expensive. Strict python libraries like [vncdotool](https://github.com/sibson/vncdotool) are for automating cli tasks, and don't have all the necessary features. Everything else I can find is either closed source and don't support python ([TightVNC sdk](https://www.tightvnc.com/licensing-dotnetviewersdk.php)), or has a small community behind it.

**Additional context**

If this library turns out to be the best route for my project, I'd be willing to take a stab at implementing this over my weekends, if it's something you're interested in.

**Comment by bk138** [comment]
Yes, that would be valuable. Also, I think that once one FFI binding is in place, others might be easier to add. So go ahead!

**Comment by Cameronsplaze** [comment]
Just an update on this, I'll try to work on it soon at https://github.com/Coldtoe-Devs/libvncserver.

I was originally researching [pybind11](https://pybind11.readthedocs.io/en/stable/), but they require you to explicitly list every bind you want. (Which means you'd have to maintain a bind list going forward, and might forget to forward a feature to python).

Thankfully, looks like there's a couple projects built on top of pybind11 to use it automatically, listed [here](https://pybind11.readthedocs.io/en/stable/compiling.html#generating-binding-code-automatically) (binder, AutoWig, and robotpy-build). I'll need to research which might be the best one to use here.

**Comment by Cameronsplaze** [comment]
It seems like pybind11 requires `c++11` to compile, so I'm not sure any of those packages built on top of it will work on a C library. I can get the farthest so far with robotpy-build. I through together a basic c++ library, and got the binds working on that. But when trying the same process here on a C lib, it won't compile. 

I'll keep researching a work around just in case, but what are your thoughts on switching to `C++` from `C`? It'd let you use [unique_ptr](https://en.cppreference.com/w/cpp/memory/unique_ptr) and [shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr) too I believe, and they're strongly recommended by all of the binding packages.

**Comment by bk138** [comment]
I'd say introducing C++ makes things overly complicated as LibVNCServer is a C library and will never be a C++ one. Using C++ for Python bindings sounds like quite an addition of code for a single purpose and thus a lot of maintenance work, but I might be wrong...

**Comment by bk138** [comment]
It would always be better to go with less dependencies, but if C++ is needed for painless Python bindings and all the Python/C++ is a CMake Opt-In, it would be OK from my side. Can you sketch out the general approach and/or what a developer's workflow would be in generating bindings?

**Comment by Cameronsplaze** [comment]
Since there's no way to compile this library in c++, I've been learning SWIG, which can generate bindings for C libraries. The nice thing is too, if you get it working for one language like python, you get a bunch of languages almost for free through it.

I have a minimal example I've been playing with here: https://github.com/Coldtoe-Devs/TestingSWIG. It also has todo's I need to learn, before I'm ready to try adding it to a library this large. Whenever you have a chance, could you review it, and see if there's any blockers or concerns for using swig here? I put any I had in the readme, with whatever else I thought was important.

**Comment by bk138** [comment]
SWIG sounds good, let me know once you have a working proof of concept.

**Comment by Cameronsplaze** [comment]
After some research, I think [spice](https://www.spice-space.org/) might work out better for my project than VNC, so I'm not sure I'll have the motivation to keep working on this feature sadly. Sorry about dropping it like this. If someone was to pick this up for ANY language bindings, I do strongly think they should give SWIG at least a look.

Would you rather keep this issue open for if someones interested down the line? Or just have them open up their own issue?

**Comment by bk138** [comment]
No problem at all, it's your time and your project! We'll leave this issue open and see if there's interest from other parties. Thanks for your effort so far!
